### ISSUE TITLE
Support querying alerts per relation in PrometheusRemoteWriteProvider

### ISSUE THREAD

**Issue by balbirthomas** [canonical/prometheus-k8s-operator#288]
### Use Case

At present the `PrometheusRemoteWriteProvider` object provides an `alerts()` method that returns all the alerts for all the relations that the provider charm has. A remote write provider charm (for example Mimir) can have a large number of related consumer charms each of which may provide a large number of alert rules. Hence a provider charm needs to be able to selectively update alert rules for a specific relation in response to a change in that relation. The currently provided `alerts()` method in the `PrometheusRemoteWriteProvider` object does not provide an API that makes this easy. It may be possible to iterate over all alert rule groups returned by `alerts()` and find the alert rule group for a specific relation by matching the group name ("Topology Identifier") against the related charm's application name. 

### Proposal

Add an optional `relation` argument to the `alerts()` method so that it only returns alert rules for that specific relation. With such an enhanced API updates to a alert rules for a specific relation may be written as follows

```
def _on_remote_write_relation_changed(self, event):
       alert_rule_groups = self._remote_write_provider.alerts(event.relation)
       self._update_alert_rules(alert_rule_groups)
```

### To Reproduce

NA

### Environment

NA

### Relevant log output

```shell
NA
```

### Additional context

NA

**Comment by rbarry82** [comment]
We [already do this in Grafana](https://github.com/canonical/grafana-k8s-operator/blob/a2db8772c929ce0d1c91cffc78052341c948f0df/lib/charms/grafana_k8s/v0/grafana_source.py#L499) (both source and dashboard). 

In some sense, it's nice because we get to precache things and don't need to query every single relation every time. It would make it easy to implement this elsewhere, since there's a pattern for it.

The downside is that the move away from `StoredState` put Grafana in a scenario where a `peer` relation was required. For Prometheus itself, that's probably _ok_. We have to be careful about not tracking this pattern into `MetricsEndpointProvider` or `PrometheusRemoteWriteConsumer`, though.

**Comment by balbirthomas** [comment]
> We have to be careful about not tracking this pattern into `MetricsEndpointProvider` or `PrometheusRemoteWriteConsumer`, though.

Yes, we can avoid breaking existing clients by making the relation  argument of `alerts()` optional.

**Comment by balbirthomas** [comment]
Also I should mention the use case described above is specific to the long term storage component of  a high availability solution because it will be aggregating metrics from many sources. For something like Prometheus this may not typically be the case. Also Prometheus' way to handling alert rules is just writing them *all* to disk. But Mimir selectively updates alert rules. Hence it is expensive to update *all* alert rules just because a subset of them changed. So while the existing API is sufficient for the present use case it may be inefficient for future use cases.

**Comment by rbarry82** [comment]
I think I'm missing something here.

The workload/charm will never be aware of a relation event (or specific relations in general), so it will never know which arg to pass. De-facto, it retrieves all alerts and flushes them all to disk.

Either an event would need to be passed to the workload indicating "hey, this relation changed", which it could pass into `alerts()` (which would also require changes to the way alerts are written out to the container, since updated alerts would need to be conditionally deleted and new ones written, using the identifier to "pick them out"), or  `relation_changed` would update a data bucket similar to Grafana which is a map of `relation.id -> alert_rules`, so at least _that_ part of it could be selectively updated, and we don't trigger a flood of `relation-get`s every time anything is written to disk.

Arbitrarily, the peer bucket makes more sense for a horizontally sharded/scaled workload anyway, since the peer bucket would be shared between them all.

Grafana is slated to add an HTTP endpoint for their own alerting in 8.5.something, so it could be shared, but Mimir doesn't appear to have any docs for the sort-of 12factor provisioning we're currently doing.

Instead, there is an [HTTP endpoint](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#set-rule-group) to update per-namespace. I doubt if we've decided what a "namespace" is going to mean in our terminology, but it's either:

* One namespace per related application. In this case, they could be individually updated.
* One namespace for COS.

Either way, our current design of:

```yaml
groups:
  name: foo
  rules: ...
```

Cannot work. Namespacing notwithstanding, which is just a top-level key anyway, the structure certainly looks like we will need to concatenate rules. Either all rules per "namespace" (if a namespace is per relation) or _all rules_ if we only use one. 

We have to rely on Mimir "doing the right thing" since the API endpoint claims that it will update, so it's hopefully smarter than guess here.

The tricky problem looks like it's going to be "how do we _remove_ rules without destroying the entire object and recreating it", which will certainly be expensive. There's [an API endpoint for it](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#delete-rule-group) and again, Grafana already does something similar. Strategies for that could be debated (keep the "old" rule definitions in peer data, generate the new one, "pick out" `namespace -> group -> name` keys to delete, and handle it selectively, or...

I'm still unclear on where an arg to `alerts()` would come in. Spitballing, this all looks like it would need to be handling backend data operations. It's arguably close enough that it could all be done in Prom, and the Mimir workload charm would be responsible for taking the data from `alerts()` and concatenating it in the charm itself, but it warrants discussion.

**Comment by balbirthomas** [comment]
> I'm still unclear on where an arg to `alerts()` would come in

Trying to clarify. At present `alerts()` returns *all alert rules* i.e. set by any related charm. The `alerts()` methods does this be iterating over all relations [here](https://github.com/canonical/prometheus-k8s-operator/blob/main/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1090) . There is a use case (as described above) to request alert rules set by a specific related charm. For this purpose providing an *optional* relation argument to the `alerts()` method may be used. If this argument is provided then the `alerts()` method can return alert rules for just that relation rather than all relations.

**Comment by rbarry82** [comment]
Let me rephrase:

The charm (well, `charm.py`, where `alerts()` is called) is abstracted away from the nitty-gritty part of relations. The only awareness it has is when an event is emitted from the provider, and that doesn't map back to a relation.

Under what circumstances would the charm "know" that it should request a refresh of alerts from some relation? Why _that_ relation? Why would we make the assumption that the data differed from what would have been in the relation databag?

Furthermore, assuming the above can be answered, the structure of alerts must be solved. One namespace per relation? Global? How does the charm know which keys to delete from the API endpoint?

Performance in the sense of "don't ask every relation every time, cache it" makes sense, but I can't quire picture how or why this argument would ever be used. Conceptually it makes sense, just not an actual call chain to get there, or what we do with partial alert rule data when we get it.
